objects/src/assets/mod.rs

```diff
@@ -151,11 +149,14 @@ impl TryFrom<Word> for Asset {
     type Error = AssetError;
 
     fn try_from(value: Word) -> Result<Self, Self::Error> {
-        let first_bit = value[3].as_int() >> 63;
-        match first_bit {
-            0 => NonFungibleAsset::try_from(value).map(Asset::from),
-            1 => FungibleAsset::try_from(value).map(Asset::from),
-            _ => unreachable!(),
+        // For fungible assets, the 4th word contains the account_id, and the bit is set to true.
+        // For non-fungible assets the account_id is on the 2nd word. The 4th word is initialized
+        // with the asset's content hash, and then the bit is set to false.
```

Comment:
The main reason is to control collisions which may happen when assets are added to account vault. There, the most significant element of the word (the element with index 3) determines if one asset will end up in the same leaf as another asset. So, with this arrangement, we achieve:

- Every fungible asset will end up in a different leaf as the most significant element of the asset word is set to faucet ID. This is good because, for fungible assets, all assets with the same faucet ID can be merged together.
- For non-fungible assets, there is a high probability that assets issued by the same faucet will end up in different leaves. This is also good as we want to minimize potential of collisions. If we had used the same rule as for fungible assets, then all assets issued by the same faucet would end up in the same leaf in the asset vault.

Reply:
Why not use the the asset's hash as its key in the vault instead? The current setup seems quite brittle tbh.

Reply:
We can't use hash for fungible assets because they include amounts and we want all fungible assets issues by the same faucet to end up in the same leaf (i.e., have the same key).

Another property we want to achieve is that by looking on the asset word, it is possible to tell which faucet issued this asset - and this is why we swap one of the elements of non-fungible assets with faucet ID.